src/devices/cpu/drcbearm64.cpp

```diff
@@ -0,0 +1,3901 @@
+	a.bfi(FLAGS_REG, SCRATCH_REG2, 4, 1);
+}
+
+void drcbe_arm64::get_unordered(a64::Assembler &a, const a64::Gp &reg) const
+{
+	get_shifted_bit(a, reg.x(), FLAGS_REG, 1, 4);
+}
+
+void drcbe_arm64::store_carry_reg(a64::Assembler &a, const a64::Gp &reg) const
+{
+	a.bfi(FLAGS_REG, reg.x(), 0, 1);
+}
+
+void drcbe_arm64::store_carry(a64::Assembler &a, bool inverted) const
+{
+	if (inverted)
+		a.cset(SCRATCH_REG1, a64::CondCode::kCC);
+	else
+		a.cset(SCRATCH_REG1, a64::CondCode::kCS);
+
+	store_carry_reg(a, SCRATCH_REG1);
+}
+
+void drcbe_arm64::get_carry(a64::Assembler &a, const a64::Gp &reg, bool inverted) const
+{
+	a.and_(reg.x(), FLAGS_REG, 1);
+
+	if (inverted)
+		a.eor(reg.x(), reg.x(), 1);
+}
+
+void drcbe_arm64::load_carry(a64::Assembler &a, bool inverted) const
+{
+	a.mrs(SCRATCH_REG1, a64::Predicate::SysReg::kNZCV);
+	a.bfi(SCRATCH_REG1, FLAGS_REG, 29, 1);
+
+	if (inverted)
+		a.eor(SCRATCH_REG1, SCRATCH_REG1, 1 << 29);
+
+	a.msr(a64::Predicate::SysReg::kNZCV, SCRATCH_REG1);
+}
+
+void drcbe_arm64::get_shifted_bit(a64::Assembler &a, const a64::Gp &dst, const a64::Gp &src, uint32_t bits, uint32_t shift) const
+{
+	a.lsr(dst.x(), src.x(), shift);
+	a.and_(dst.x(), dst.x(), bits);
+}
+
+void drcbe_arm64::calculate_carry_shift_left(a64::Assembler &a, const a64::Gp &reg, const a64::Gp &shift, int maxBits) const
+{
+	Label skip = a.newLabel();
+	a.cbz(shift, skip);
+
+	const a64::Gp scratch = select_register(SCRATCH_REG1, reg.isGpW() ? 4 : 8);
+	const a64::Gp scratch2 = select_register(SCRATCH_REG2, reg.isGpW() ? 4 : 8);
+
+	// carry = ((PARAM1 << (s	hift - 1)) >> maxBits) & 1
+	a.sub(scratch, shift, 1);
+	a.lsl(scratch2, reg, scratch);
+	a.lsr(scratch2, scratch2, maxBits);
+	store_carry_reg(a, scratch2);
+
+	a.bind(skip);
+}
+
+void drcbe_arm64::calculate_carry_shift_left_imm(a64::Assembler &a, const a64::Gp &reg, const int shift, int maxBits) const
+{
+	if (shift == 0)
+		return;
+
+	const a64::Gp scratch = select_register(SCRATCH_REG1, reg.isGpW() ? 4 : 8);
+
+	// carry = ((PARAM1 << (shift - 1)) >> maxBits) & 1
+	a.lsl(scratch, reg, shift - 1);
+	a.lsr(scratch, scratch, maxBits);
+	store_carry_reg(a, scratch);
+}
+
+void drcbe_arm64::calculate_carry_shift_right(a64::Assembler &a, const a64::Gp &reg, const a64::Gp &shift) const
+{
+	Label skip = a.newLabel();
+	a.cbz(shift, skip);
+
+	const a64::Gp scratch = select_register(SCRATCH_REG1, reg.isGpW() ? 4 : 8);
+	const a64::Gp scratch2 = select_register(SCRATCH_REG2, reg.isGpW() ? 4 : 8);
+
+	// carry = (PARAM1 >> (shift - 1)) & 1
+	a.sub(scratch, shift, 1);
+	a.lsr(scratch2, reg, scratch);
+	store_carry_reg(a, scratch2.x());
+
+	a.bind(skip);
+}
+
+void drcbe_arm64::calculate_carry_shift_right_imm(a64::Assembler &a, const a64::Gp &reg, const int shift) const
+{
+	if (shift == 0)
+		return;
+
+	const a64::Gp scratch = select_register(SCRATCH_REG1, reg.isGpW() ? 4 : 8);
+
+	// carry = (PARAM1 >> (shift - 1)) & 1
+	a.lsr(scratch, reg, shift - 1);
+	store_carry_reg(a, scratch.x());
```

Comment:
Same here, why the difference between left and right shifts with the argument to `store_carry_reg`?

Reply:
Also no reason for this one (see above comment).